theano/tensor/opt.py

```diff
@@ -221,89 +221,124 @@ def inplace_elemwise_optimizer(fgraph):
-
-            for candidate_output in candidate_outputs:
-                for candidate_input in candidate_inputs:
-                    # remove inputs that don't have the same dtype as the output
-                    if node.inputs[candidate_input].type != node.outputs[
-                        candidate_output].type:
-                        continue
-
-                    inplace_pattern = dict(baseline)
-                    inplace_pattern[candidate_output] = candidate_input
-                    try:
-                        if hasattr(op.scalar_op, "make_new_inplace"):
-                            new_scal = op.scalar_op.make_new_inplace(
-                                scalar.transfer_type(
-                                    *[inplace_pattern.get(i, None) \
-                                          for i in xrange(len(node.outputs))]))
-                        else:
-                            new_scal = op.scalar_op.__class__(
-                                scalar.transfer_type(
-                                    *[inplace_pattern.get(i, None) \
-                                          for i in xrange(len(node.outputs))]))
-                        new_outputs = OP(new_scal, inplace_pattern)(
+        # gpuarray GpuElemwise inherit from Elemwise
+        elemwise_nodelist = [node for node in
+                             list(graph.io_toposort(fgraph.inputs,
+                                                    fgraph.outputs))
+                             if type(node.op) == OP]
+        protected_inputs = [
+            f.protected for f in fgraph._features if
+            isinstance(f, theano.compile.function_module.Supervisor)]
+        protected_inputs = sum(protected_inputs, [])  # flatten the list
+        protected_inputs.extend(fgraph.outputs)
+
+        # See: https://github.com/Theano/Theano/issues/2962
+        if len(elemwise_nodelist) > 0:
+            check_each_change = int(numpy.ceil(numpy.log(
+                len(elemwise_nodelist))))
+            check_each_change = max(check_each_change, 1)
+        else:
+            return
+        failed_list = set()
+        prev_len_elemlist = len(elemwise_nodelist) + 1
+        changed_node_since_validate = set()
+        while check_each_change > 0 and prev_len_elemlist > len(elemwise_nodelist):
+            prev_len_elemlist = len(elemwise_nodelist)
+            for node in elemwise_nodelist:
+                op = node.op
+                baseline = op.inplace_pattern
+                candidate_outputs = [i for i in xrange(len(node.outputs))
+                                     if i not in baseline]
+                # node inputs that are Constant, already destroyed,
+                # fgraph protected inputs and fgraph outputs can't be used as
+                # inplace target.
+                # Remove here as faster.
+                candidate_inputs = [i for i in xrange(len(node.inputs))
+                                    if i not in baseline.values() \
```

Comment:
the "\" at the end of the line are not pep8 compliant and in this case, are not useful. Just remote them.